### ISSUE TITLE
[Bug] SM120 grouped FP8 DeepGEMM weight preparation skips UE8M0 requantization

### ISSUE THREAD

**Issue by number-eleven11** [sgl-project/sglang#39063]
# [Bug] SM120 grouped FP8 DeepGEMM weight preparation skips UE8M0 requantization

## Checklist

- [x] I searched related issues but found no solution for this specific guard.
- [x] Reproduced with the latest SGLang main checked out for this test.
- [x] Environment and a minimal reproduction are provided.
- [x] This is a bug report, written in English.

## Describe the bug

On NVIDIA RTX PRO 5000 72GB (SM120), SGLang enables DeepGEMM and UE8M0 scales when a compatible DeepGEMM build is installed. However, `should_deepgemm_weight_requant_ue8m0` unconditionally returns False for SM120, including grouped MoE weights.

`Fp8MoEMethod.process_weights_after_loading` calls `requant_block_scale_ue8m0_for_deepgemm` for w13/w2, but that helper therefore skips conversion. Passing non-power-of-two FP32 block scales to the grouped FP8 DeepGEMM path then triggers its scale-layout assertion.

This report is based on a **single-GPU weight-preparation + grouped-GEMM reproduction**, not a full-model startup or accuracy run. The motivating workload is GLM-5.3-Flash blockwise FP8 MoE. No checkpoint download is needed for the reproduction.

## Versions

- SGLang main: `ad5af539cd39e9af10ef318d3f8a50f5d6f9c153` (fresh checkout, production source unmodified).
- DeepGEMM nv_dev: `572557e7ae9ad5331b81a1c250f141fba2c57962`, package `2.6.1+local` with SM120 grouped FP8 support.
- GPU: NVIDIA RTX PRO 5000 72GB Blackwell, capability `(12, 0)`.
- PyTorch: `2.13.0+cu130`; CUDA runtime: `13.0`.

## Reproduction

Run in an environment with the above SGLang and DeepGEMM installed:

```bash
SGLANG_ENABLE_JIT_DEEPGEMM=1 python repro.py
# Run the control in a NEW process after the device assertion:
SGLANG_ENABLE_JIT_DEEPGEMM=1 REPRO_MODE=explicit_requant_control python repro.py
```

```python
import os
import torch
import sglang
import deep_gemm as dg
from sglang.srt.layers import deep_gemm_wrapper as wrapper
from sglang.srt.layers.quantization.fp8_utils import (
    requant_block_scale_ue8m0_for_deepgemm,
    requant_weight_ue8m0_inplace,
)

print("SGLang source:", sglang.__file__)
print("DeepGEMM source:", dg.__file__)
torch.manual_seed(123)
w = torch.nn.Parameter(
    torch.randn(2, 256, 256, device="cuda").to(torch.float8_e4m3fn),
    requires_grad=False,
)
ws = torch.nn.Parameter(
    torch.full((2, 2, 2), 0.03, device="cuda", dtype=torch.float32),
    requires_grad=False,
)
print(wrapper.ENABLE_JIT_DEEPGEMM, wrapper.DEEPGEMM_SCALE_UE8M0)
print(requant_block_scale_ue8m0_for_deepgemm(
    w, ws, [128, 128], use_deepgemm_runner=True,
    output_dtype=torch.bfloat16, weight_shape=(256, 256),
))
q = torch.randn(256, 256, device="cuda").to(torch.float8_e4m3fn)
qs = torch.full((256, 2), 0.03125, device="cuda", dtype=torch.float32)
idx = torch.arange(2, device="cuda", dtype=torch.int32).repeat_interleave(128)
out = torch.empty(256, 256, device="cuda", dtype=torch.bfloat16)
if os.environ.get("REPRO_MODE") == "explicit_requant_control":
    requant_weight_ue8m0_inplace(w, ws, [128, 128])
dg.m_grouped_fp8_gemm_nt_contiguous((q, qs), (w, ws), out, idx)
torch.cuda.synchronize()
print("finite:", torch.isfinite(out).all().item())
```

## Actual results

Unmodified upstream weight-preparation path:

```text
FLAGS True True
GUARD False
PREPARE False
Assertion failed: deep_gemm/impls/smxx_layout.cuh:131,
condition: (values[j] & 0x807fffffu) == 0
CUDA_ERROR_LAUNCH_FAILED
```

Separate-process control explicitly calling the existing requantization helper:

```text
FLAGS True True
GUARD False
PREPARE False
MODE explicit_requant_control
PASS finite= True
```

The control establishes successful execution and finite outputs only; it is not an end-to-end numerical accuracy claim.

## Expected behavior / suggested direction

Preserve the SM120 dense Linear guard for CUTLASS/Triton consumers, but allow supported grouped MoE weights to follow the existing UE8M0 requantization path when DeepGEMM is selected. Merely removing the SM120 guard globally would risk converting scales for dense kernels that still expect FP32 scales.

Relevant pinned sources:

- [Capability detection](https://github.com/sgl-project/sglang/blob/ad5af539cd39e9af10ef318d3f8a50f5d6f9c153/python/sglang/srt/layers/deep_gemm_wrapper/configurer.py)
- [Unconditional SM120 guard](https://github.com/sgl-project/sglang/blob/ad5af539cd39e9af10ef318d3f8a50f5d6f9c153/python/sglang/srt/model_loader/utils.py#L265)
- [MoE weight preparation](https://github.com/sgl-project/sglang/blob/ad5af539cd39e9af10ef318d3f8a50f5d6f9c153/python/sglang/srt/layers/quantization/fp8.py#L1809)

## Environment

`check_env` reports installed distribution metadata for SGLang (`0.0.0.dev1+gf609d677b`), not the source override. The reproduction prints `sglang.__file__` and imports the fresh upstream checkout at the SHA above through PYTHONPATH. No older SGLang source directory is on the reproduction PYTHONPATH.

```text
Python: 3.12.3 (main, Jul 15 2026, 23:46:41) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA RTX PRO 5000 72GB Blackwell
GPU 0,1,2,3,4,5,6,7 Compute Capability: 12.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 580.126.09
PyTorch: 2.13.0+cu130
sglang: 0.0.0.dev1+gf609d677b
sglang-kernel: 0.4.6.post1
flashinfer_python: 0.6.17
flashinfer_cubin: 0.6.17
flashinfer_jit_cache: 0.6.17+cu130
triton: 3.7.1
transformers: 5.16.0.dev0
torchao: Module Not Found
numpy: 2.3.5
aiohttp: 3.14.3
fastapi: 0.141.1
huggingface_hub: 1.28.0
interegular: 0.3.3
modelscope: 1.39.1
orjson: 3.12.0
outlines: 0.1.11
packaging: 26.3
psutil: 7.2.2
pydantic: 2.13.4
python-multipart: 0.0.32
pyzmq: 27.2.0
uvicorn: 0.52.4
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.14.0
anthropic: 1.0.0
litellm: Module Not Found
torchcodec: 0.15.0+cu130
ulimit soft: 1048576
```

Topology: GPUs 0–3 share NUMA node 0 (CPU 0–63,128–191); GPUs 4–7 share NUMA node 1 (CPU 64–127,192–255). Within each group, GPU connections are NODE; between groups, SYS. This reproduction uses only GPU 0.

**Comment by liaoruoxue** [comment]
Full-model reproduction + fix verification on the same hardware/model, plus a minimal three-file patch.

**Environment** (same GPU family and model as the report): RTX PRO 5000 72GB Blackwell (SM120), GLM-5.3-Flash blockwise-FP8 MoE (288 experts, per-rank N=512/K=4096 and N=4096/K=256 at MoE-TP8/EP1), sglang 0.5.17.dev900+, DeepGEMM 0.1.5.post3 (wheel), torch 2.13.0+cu130.

**Full-model behavior**: with `--moe-runner-backend deep_gemm`, the server crashes during decode CUDA-graph capture with `CUDA_ERROR_LAUNCH_FAILED (719)`. Capturing the actual kernel arguments (one-shot dump of `lhs/rhs/out/m_indices` right before the crash) and replaying them offline reproduces the device assertion `smxx_layout.cuh:131: (values[j] & 0x807fffff) == 0` — i.e. the ue8m0 cast path rejects non-power-of-two fp32 weight scales, exactly as diagnosed above.

**Fix** (minimal, three files): thread an `is_grouped` flag through the guard so SM120 only skips requant for dense layers, not for grouped MoE:

```diff
# srt/model_loader/utils.py
-def should_deepgemm_weight_requant_ue8m0(weight_block_size, output_dtype=None, weight_shape=None):
+def should_deepgemm_weight_requant_ue8m0(weight_block_size, output_dtype=None, weight_shape=None, is_grouped=False):
     ...
-    if get_device_sm() == 120:
+    if get_device_sm() == 120 and not is_grouped:
         return False

# srt/layers/quantization/fp8.py (Fp8MoEMethod.process_weights_after_loading, w13/w2 loop)
     requant_block_scale_ue8m0_for_deepgemm(..., weight_shape=weight.shape[-2:],
+        is_grouped=True)

# srt/layers/quantization/fp8_utils.py (requant_block_scale_ue8m0_for_deepgemm signature + guard call)
+    is_grouped: bool = False
     ... pass through to should_deepgemm_weight_requant_ue8m0(..., is_grouped=is_grouped)
```

**Verification after the patch**: weight requant happens at load; decode CUDA-graph capture completes; a 30-request quality probe passes 30/30 with speculative accept_len 2.99 (identical to the triton-runner baseline), i.e. the ue8m0 requantization itself is lossless for this checkpoint.

**Note for anyone landing this**: with the requant gap fixed, execution then proceeds further and hits a second, separate issue — the SM120 grouped kernel itself is not CUDA-graph-replay safe (illegal memory access at `graph.replay()` while eager runs of the same shapes are clean). We're filing that one upstream at DeepGEMM with the evidence chain. So this fix is necessary but not yet sufficient for production `deep_gemm` on SM120.
